### PR TITLE
fix: pod finalizer removal and odd pod status

### PR DESCRIPTION
**Pull request by Joibel**
WIP

#### Part 1

If pod finalizers are in use they should not prevent pod deletion after the pod is complete.

For example: If you have a podGC.strategy of `OnPodSuccess` with a `deleteDelayDuration` set and you delete the owning Workflow during the deleteDelayDuration then the pod will remain until `deleteDelayDuration` expires. If the workflow-controller is restarted during this window the pod is orphaned, with the finalizer still in place.

`blockOwnerDeletion` in the `ownerReference` of a pod does not prevent the owner (Workflow) being deleted in all circumstances

#### Part 2

It is possible for a node to disappear from a cluster as a surprise. In this case the Pod ContainerStatus could remain in running (because the container never went into any further state), whilst the Pod's own Status is in Error.

This PR attempts to recognise this case and set the Workflow Node status accordingly.

#### Split this PR

This PR should probably be two PRs for each of these, but for now I'm making this to allow full CI to check it over.

### Verification

Updated unit test for Part 1

### CODE REVIEW

**Review by isubasinghe — APPROVED** [review]
This makes sense, I'm not entirely confident every edge case is accounted for but lets ship it for now.

**Comment by Joibel** [comment]
@shuangkun and @jswxstw, I'd like to hear your thoughts on these changes if you have time.

**Comment by Joibel** [comment]
I realise now that #13491 does something similar (I originally wrote this patch against the 3.5 codebase).

In the scenario we're seeing with a non-evicted pod, the workflow node still gets stuck with #13491 in place.

File: workflow/controller/pod_cleanup.go [hunk 1]
```diff
@@ -37,9 +39,10 @@ func (woc *wfOperationCtx) queuePodsForCleanup() {
 		if !nodePhase.Fulfilled() {
 			continue
 		}
-		switch determinePodCleanupAction(selector, pod.Labels, strategy, workflowPhase, pod.Status.Phase) {
+		switch determinePodCleanupAction(selector, pod.Labels, strategy, workflowPhase, pod.Status.Phase, pod.Finalizers) {
 		case deletePod:
 			woc.controller.queuePodForCleanupAfter(pod.Namespace, pod.Name, deletePod, delay)
+		case removeFinalizer:
```
Comment (shuangkun) [inline workflow/controller/pod_cleanup.go:45]: Reasonable, is it need to add 
```
wfc.queuePodForCleanup(p.Namespace, p.Name, removeFinalizer) 
```
here？

Comment (Joibel) [reply workflow/controller/pod_cleanup.go:45]: Yeah, oops. I had thought the final parameter was `action`. Will fix.

Comment (isubasinghe) [reply workflow/controller/pod_cleanup.go:45]: Nice catch @shuangkun, I missed this in the review.

Comment (isubasinghe) [reply workflow/controller/pod_cleanup.go:45]: Maybe change this section to this: 
```go
		action := determinePodCleanupAction(selector, pod.Labels, strategy, workflowPhase, pod.Status.Phase, pod.Finalizers)
		if action == deletePod {
			woc.controller.queuePodForCleanupAfter(pod.Namespace, pod.Name, action, delay)
		} else {
			woc.controller.queuePodForCleanup(pod.Namespace, pod.Name, action)
		}
```

**Comment by jswxstw** [comment]
> It is possible for a node to disappear from a cluster as a surprise. In this case the Pod ContainerStatus could remain in running (because the container never went into any further state), whilst the Pod's own Status is in Error. We have seen this in real clusters, but it is rare.

Indeed, this issue https://github.com/kubernetes/kubernetes/issues/98718 in older versions of Kubernetes can also cause inconsistencies between pod status and container status, which can lead to the workflow getting stuck.
